Model Loader, Fixes

d8ahazard opened the pull request:
For the sake of sanity, ever scaler/model I'm referring to below will just be called "scalers".

Add universal model loader:
All scalers that require a model to be downloaded have at least one default URL so the user doesn't have to grab/add anything.
All scalers that download models will save models to their respective directory in /models, and create their own folders as needed.
All scalers that load/fetch models will not do so until they are needed during runtime.
All scalers have an option to override/append the default model path at runtime.
Check for existing models in "old" paths and move them to the new ones.

Fix various typos and optimize imports, etc, as I was going through files.

Add an arch file for gfpgan_model so we can also download weights to the /models directory.

d8ahazard committed:
Fix swinir arch name

d8ahazard committed:
Rename swinir -> swinir_model

d8ahazard committed:
Re-implement universal model loading

d8ahazard committed:
Cleanup existing directories, fixes

d8ahazard committed:
Update .gitignore

d8ahazard committed:
Cleanup gitignore, old files

AUTOMATIC1111 commented:
If you just want to change paths for gfpgan you can do that by just assigning to modules' functions, do not copy the whole file just for this.

Do not shutil.rmtree directories that are not empty.

if you're placing models each into own directory, same should be done with stable diffusion models. Right now it works I think because no other kind of model is using .ckpt extension, but if any appear, the program would list and try to load them as stable diffusion models. `stable-diffusion` subdir for stable diffusion models should work.

d8ahazard committed:
Use model loader with stable-diffusion too.

Hook the model loader into the SD_models file.
Add default url/download if checkpoint is not found.
Add matching stablediffusion-models-path argument.
Add message that --ckpt-dir will be removed in the future, but have it pipe to stablediffusion-models-path for now.
Update help strings for models-path args so they're more or less uniform.
Move sd_model "setup" call to webUI with the others.
Ensure "cleanup_models" method moves existing models to the new locations, including SD, and that we aren't deleting folders that still have stuff in them.

d8ahazard committed:
Merge remote-tracking branch 'upstream/master' into ModelLoader

d8ahazard commented:
> If you just want to change paths for gfpgan you can do that by just assigning to modules' functions, do not copy the whole file just for this.
> 
> Do not shutil.rmtree directories that are not empty.
> 
> if you're placing models each into own directory, same should be done with stable diffusion models. Right now it works I think because no other kind of model is using .ckpt extension, but if any appear, the program would list and try to load them as stable diffusion models. `stable-diffusion` subdir for stable diffusion models should work.

OK, SD has the model loader hooked in now. Ensured folders are empty before deletion as well.

Comment on including GFPGAN is above.

d8ahazard committed:
Remove debugging message

AUTOMATIC1111 commented:
> If you know a trick to inject this into the download_from_url function, I'd love to know it.

Something like this should work:
```
gfpgan.utils.load_file_from_url_orig = gfpgan.utils.load_file_from_url

def my_load_file_from_url(**kwargs):
    kwargs["url"] = "..."
    return load_file_from_url_orig(kwargs)

gfpgan.utils.load_file_from_url = my_load_file_from_url
```

Same with gfpgan.utils.FaceRestoreHelper.

d8ahazard commented:
> load_file_from_url_orig = gfpgan.utils.load_file_from_url
> 
> def my_load_file_from_url(**kwargs):
>     kwargs["url"] = "..."
>     return load_file_from_url_orig(kwargs)
> 
> gfpgan.utils.load_file_from_url = my_load_file_from_url

Problem I'm having is that I want to change the destination folder, not the source URL for the files. I just want them all to go to /models/GFPGAN. 

Any help would be greatly appreciated. I've got a few other PR's about ready to go as well, and I'd love to stop deleting and re-downloading my models every time I switch back to master. :P

AUTOMATIC1111 commented:
```python
load_file_from_url_orig = gfpgan.utils.load_file_from_url

def my_load_file_from_url(**kwargs):
    kwargs["model_dir"] = "..."
    return load_file_from_url_orig(kwargs)

gfpgan.utils.load_file_from_url = my_load_file_from_url
```

d8ahazard committed:
Holy $hit.

Yep.

Fix gfpgan_model_arch requirement(s).
Add Upscaler base class, move from images.
Add a lot of methods to Upscaler.
Re-work all the child upscalers to be proper classes.
Add BSRGAN scaler.
Add ldsr_model_arch class, removing the dependency for another repo that just uses regular latent-diffusion stuff.
Add one universal method that will always find and load new upscaler models without having to add new "setup_model" calls. Still need to add command line params, but that could probably be automated.
Add a "self.scale" property to all Upscalers so the scalers themselves can do "things" in response to the requested upscaling size.
Ensure LDSR doesn't get stuck in a longer loop of "upscale/downscale/upscale" as we try to reach the target upscale size.
Add typehints for IDE sanity.
PEP-8 improvements.
Moar.

d8ahazard committed:
Merge remote-tracking branch 'upstream/master' into ModelLoader

d8ahazard committed:
Fixes, cleanup.

d8ahazard commented:
@AUTOMATIC1111 
@C43H66N12O12S2 

So, yeah. I did a thing. Or, a lot of things. You're probably going to hate me, or this at the very least.

The modelloader file now dynamically loads all upscalers just by virtue of being a subclass of the Upscaler class. This includes dynamically loading the requisite command-line args, if present.

Fixed the Gfpgan stuff and hooked the directories properly.

And the big part - I re-jiggered all of the upscaler files so they're proper classes that all do the same things and have a few backflips to do so in the process. Most of the code is still the same, I just moved some of it around to fit the class structure, and adjusted some of the methods to be more universal in the process.

Also added BSRGAN as a scaling option. 

Shortened up names of Real-ESRGAN scalers so they don't take up half the page, made the names of all the scalers more uniform.

LDSR now no longer requires a separate git repository - I've cleaned up the single class file that's taken from the base notebookhelpery.py in the SD project and adjusted it so it properly scales to adapt to the requested upscaling size, without having to run 3x times in the scaler going up and down before reaching the desired height. Sometimes, it's slower, but it only runs once, and it runs right.

I know it's a lot to review, but please take the time to test it and let me know what you think. I'd like to also look at adding a tiling option for all upscalers (goBig?), but just want to get this merged to master so I don't have to keep fixing conflicts. ;)

Sorry bout the mess. I didn't refactor anything, just organized the files so they were easier to duplicate/edit in the future.

AUTOMATIC1111 commented:
are you actively making an effort to make it as difficult to review and merge as possible

AUTOMATIC1111 commented:
Thank you for those changes. I do like some of them, like organizing models and controlling where downloads go.

I absolutely hate the way this was sumbitted and the way this was tested. 2552204fcb04c13af93749bd125d086ba148517c a5e7b371d61bc063062b636236f1f37c264f4115 8f1b3153180da3b626529a7d077eac63d4b2e5e4. Those errors could have been detected if the changes were small and isolated. I made a https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing section to prevent this in future and I will without blinking reject a PR of this size if comes unannounced from now on.

MiguelRita commented:
This one killed most of the SD upscaling in the script and the Extras tab.
Also in the Extra tab after the second run, there is permission error:

PermissionError: [Errno 13] Permission denied: 'F:\StableDiffusion\\stable-diffusion-webui-1111\\models\\RealESRGAN'

AUTOMATIC1111 commented:
#1354

d8ahazard commented:
> Thank you for those changes. I do like some of them, like organizing models and controlling where downloads go.
> 
> I absolutely hate the way this was sumbitted and the way this was tested. [2552204](https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/2552204fcb04c13af93749bd125d086ba148517c) [a5e7b37](https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/a5e7b371d61bc063062b636236f1f37c264f4115) [8f1b315](https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/8f1b3153180da3b626529a7d077eac63d4b2e5e4). Those errors could have been detected if the changes were small and isolated. I made a https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing section to prevent this in future and I will without blinking reject a PR of this size if comes unannounced from now on.

Sorry. I try to keep things small, but this one sort of snowballed due to the over-arching nature of the changes I was making. 

But, I think that with the big stuff now being part of main, any subsequent updates I may have in mind should be less impactful.
